QModManager/Patcher.cs

```diff
@@ -340,6 +342,23 @@ internal static void RemoveDuplicateModIDs()
             }
         }
 
+        internal static Dictionary<QMod, List<MethodInfo>> GetMessageRecievers(Assembly assembly)
+        {
+            return assembly.GetTypes()
+                           .SelectMany(t => t.GetMethods())
+                           .Where(m => m.IsStatic && m.GetParameters().Length == 3)
+                           .Where(m => m.GetCustomAttributes(typeof(MessageReceiver), false).Length == 1)
+                           .GroupBy(m => ((MessageReceiver)m.GetCustomAttributes(typeof(MessageReceiver), false)[0]).Sender)
+                           .Select(g => new KeyValuePair<QMod, List<MethodInfo>>(g.Key, g.ToList()))
+                           .Add(new KeyValuePair<QMod, List<MethodInfo>>(null,
+                   assembly.GetTypes()
+                           .SelectMany(t => t.GetMethods())
+                           .Where(m => m.IsStatic && m.GetParameters().Length == 3)
+                           .Where(m => m.GetCustomAttributes(typeof(GlobalMessageReciever), false).Length == 1)
+                           .ToList()))
+                           .ToDictionary(k => k.Key, v => v.Value);
```

Comment:
All these constraints make sense, but they also look like they could be greatly simplified by the use of an `interface`.  
You wouldn't be able to use `static` methods for this anymore, and modders may need to implement their own singleton patterns for it, but it would make it very easy to inform modders what the correct method signature is for these features.

This is only a suggestion in case you haven't thought of it yet.

Reply:
I dont think this would work well, since the attribute would still be needed to define the sender. Of course, this can by bypassed by sending it in the constructor, but I'm not sure if that would be the best approach

Reply:
Commit 2c4fcc2610bc408: replaced attributes with abstract classes which enforce the constraints